### PR TITLE
feat(auth): Overriding sign in when the State machine is already in the signing in state

### PR DESCRIPTION
**Pull request by gpanshu**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:* Currently if the user is already in the signing in state and a new sign in requested event takes place it is ignored. This will fix that. 

*How did you test these changes?*
(Please add a line here how the changes were tested)

- [x] Added Unit Tests
- [x] Added Integration Tests

*Documentation update required?*
- [x] No
- [ ] Yes (Please include a PR link for the documentation update)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 1]
```diff
@@ -469,14 +469,37 @@ internal class RealAWSCognitoAuthPlugin(
...
                 is AuthenticationState.SignedIn -> onError.accept(SignedInException())
+                is AuthenticationState.SigningIn -> {
+                    val token = StateChangeListenerToken()
+                    authStateMachine.listen(
+                        token,
+                        { authState ->
+                            when (authState.authNState) {
+                                is AuthenticationState.SignedOut -> {
+                                    authStateMachine.cancel(token)
+                                    val signInOptions = options as?
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:489]: Please lift up the options to the top of the function.

File: aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt [hunk 2]
```diff
@@ -469,14 +469,37 @@ internal class RealAWSCognitoAuthPlugin(
...
+                                        AWSCognitoAuthSignInOptions ?: AWSCognitoAuthSignInOptions.builder()
+                                        .authFlowType(configuration.authFlowType)
+                                        .build()
+                                    _signIn(username, password, signInOptions, onSuccess, onError)
+                                }
+                                else -> Unit
+                            }
+                        },
+                        {
+                            authStateMachine.send(AuthenticationEvent(AuthenticationEvent.EventType.CancelSignIn()))
```
Comment (div5yesh) [inline aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/RealAWSCognitoAuthPlugin.kt:499]: 🎉

**Review by div5yesh — DISMISSED** [review]
nit, else LGTM

**Review by tylerjroach — COMMENTED** [review]
Good once Divyesh's comment is implemented. Eventually it would be nice to implement this method using coroutines to remove the # of callbacks we have here.

**Review by div5yesh — APPROVED** [review]

**Review by tylerjroach — APPROVED** [review]

**Pull request merged**